### ISSUE TITLE
[Discussion] Some new API I'd like

### ISSUE THREAD

**Issue by MrIvanPlays** [PaperMC/Velocity#449]
I want to discuss some things before opening a PR.

I'd like an api where you'd be able to create a connection request with the ServerInfo object. ( Player#createConnectionRequest(ServerInfo) )
The reasoning behind is that it is less heavy to store ServerInfo objects rather than RegisteredServer objects.

The other API is about KickedFromServerEvent . I'd like a way to override the "SUCCESS" message ( or [this](https://github.com/VelocityPowered/Velocity/blob/dev/1.1.0/proxy/src/main/java/com/velocitypowered/proxy/connection/client/ConnectedPlayer.java#L697-L698) )

**Comment by LucaDev** [comment]
This would also enable the option to connect to servers without the need to register them (you need a VelocityRegisteredServer for a connection, which is not included in the API), which would in return allow using a dynamic pool of proxied servers - for example by using a kubernetes cluster as an alternative backend.

**Comment by DeprecatedLuke** [comment]
@LucaDev The api already allows you to dynamically register servers.
```
public class ServerHandler {
	private static ConcurrentHashMap<String, RegisteredServer> infoMap = new ConcurrentHashMap<>(); // basically just an ip cache
	private static ProxyServer server;

	public static void send(Player player, String ip, String name) {
		RegisteredServer info = getServerInfo(ip, name);
		player.createConnectionRequest(info);
	}

	public static RegisteredServer getServerInfo(InetSocketAddress ip, String name) {
		RegisteredServer info = infoMap.get(name);
		if (info == null) {
			info = server.registerServer(new ServerInfo(name, ip));
			infoMap.put(name, info);
		}
		return info;
	}

	public static RegisteredServer getServerInfo(String ip, String name) {
		RegisteredServer info = infoMap.get(ip.toLowerCase());
		if (info == null) {
			String[] ipArgs = ip.split(":");
			info = server.registerServer(new ServerInfo(name, new InetSocketAddress(ipArgs[0], Integer.parseInt(ipArgs[1]))));
			infoMap.put(ip, info);
		}
		return info;
	}
}
```

**Comment by Xernium** [comment]
Let me clear up some confusion here;
For your first request:
What you are requesting would be a convenience method at best. By allowing the proxy to connect players to a ServerInfo instance it would either create a registered server or you’d have to scrap the entire logic of registered servers to allow this to work. That would defeat the purpose of this being trackable through the proxy and other plugins. Registered servers are a transistent trait to the proxy server instance, and this is not going to change.   
Likewise, RegisterdServer instances may be heavier to store, but plugins don’t store that data. The data is immutable and you’re only storing a reference to a transistent object trait of the proxyserver.
About the second part of your request:
This may be available to modify in the future, but it isn’t planned for right now.

**Comment by LucaDev** [comment]
> @LucaDev The api already allows you to dynamically register servers.
> 
> [...]
> info = server.registerServer(new ServerInfo(name, ip));
> infoMap.put(name, info);
> [...]

You're right. Still: You have to [register](https://github.com/VelocityPowered/Velocity/blob/b14e7e651eab8111da58bbc76b103247d1541fcd/proxy/src/main/java/com/velocitypowered/proxy/server/ServerMap.java#L46) the server tho, which adds them into the ServerMap. If you have a high server fluctuation, you will end up with lots of servers in your list. Of course - you could always unregister the server in the end but in a highly dynamic system, I would prefer to just create a ServerInfo/RegisteredServer and send the player on there withouth having to handle the removal afterwards.

Like BungeeCords way:
```java
ProxyServer:
ServerInfo constructServerInfo(String name, InetSocketAddress address, String motd, boolean restricted);

ProxiedPlayer:
void connect(ServerInfo target);
```

**Comment by DeprecatedLuke** [comment]
You're also using java and even then, having thousands of servers won't use up a even megabyte.

**Comment by Xernium** [comment]
> > @LucaDev The api already allows you to dynamically register servers.
> > [...]
> > info = server.registerServer(new ServerInfo(name, ip));
> > infoMap.put(name, info);
> > [...]
> 
> You're right. Still: You have to [register](https://github.com/VelocityPowered/Velocity/blob/b14e7e651eab8111da58bbc76b103247d1541fcd/proxy/src/main/java/com/velocitypowered/proxy/server/ServerMap.java#L46) the server tho, which adds them into the ServerMap. If you have a high server fluctuation, you will end up with lots of servers in your list. Of course - you could always unregister the server in the end but in a highly dynamic system, I would prefer to just create a ServerInfo/RegisteredServer and send the player on there withouth having to handle the removal afterwards.

I get where you're coming from, but Velocity will not incorporate this design change for the reasons I have listed. Not only would it break API but also consistency; you're removing the ability from plugins to interact correctly with the proxy.
You are free to modify and use Velocity under the licenses of the appropriate Velocity parts, but we will most likely not pull this kind of change if proposed.

**Comment by CoreyShupe** [comment]
@MrIvanPlays This has been implemented #513

**Comment by MrIvanPlays** [comment]
not entirely tho, the 2nd part of it isn’t.
